File: Java/workspace/src/LinkedLists.java
import java.util.*;

public class LinkedLists {

    static class Node{
        public Node next = null;
        public int data;

        public Node(int d){
            data = d;
        }
    }

    //space O(n), time O(n)
    static void solution1a(Node root){
        Set<Integer> dataSet = new HashSet<>();
        Node previousNode = null;

        while(root != null){
            if(dataSet.contains(root.data)){
                previousNode.next = root.next;
            }
            else{
                dataSet.add(root.data);
                previousNode = root;
            }
            root = root.next;
        }
    }

    //space O(1), time O(n^2)
    static void solution1b(Node root){
        if(root == null || root.next == null)
            return;

        while(root != null){

            Node previousNode = root;
            Node runnerNode = root.next;
            while(runnerNode != null){
                if(root.data == runnerNode.data){
                    previousNode.next = runnerNode.next;
                }else{
                    previousNode = runnerNode;
                }
                runnerNode = runnerNode.next;
            }
            root = root.next;
        }
    }

    //space O(1), time O(n)
    static Node solution2a(Node root, int k){
        int counter = 0;

        Node current = root;
        while(current != null){
            counter++;
            current = current.next;
        }

        int kthLast = counter - k;
        if(kthLast < 0)
            return null;

        current = root;
        while(kthLast > 0){
            kthLast--;
            current = current.next;
        }

        return current;
    }

    //space O(1), time O(n), without list size
    static Node solution2b(Node root, int k){
        if(root == null || k <= 0)
            return null;

        Node runner = root;

        for(int i = 0; i < k; i++){
            if(runner == null){
                return null;
            }
            runner = runner.next;
        }

        Node current = root;
        while(runner != null){
            runner = runner.next;
            current = current.next;
        }
        return current;
    }

    //space O(1), time O(1)
    static void solution3a(Node middle){
        if(middle == null || middle.next == null)
            return;
        middle.data = middle.next.data;
        middle.next = middle.next.next;
    }

    //space O(1), time O(n)
    static Node solution4a(Node root, int x){
        Node leftPartition = null;
        Node rightPartition = null;
        Node lastLeft = null;

        while(root != null){
            Node next = root.next;
            if(root.data < x){
                if(leftPartition == null){
                    leftPartition = root;
                    lastLeft = leftPartition;
                }else{
                    lastLeft.next = root;
                    lastLeft = root;
                }
            }
            else{
                root.next = rightPartition;
                rightPartition = root;
            }
            root = next;
        }

        lastLeft.next = rightPartition;

        return leftPartition;
    }

    //space O(n), time O(n), without destroying input lists
    static Node solution5a(Node first, Node second){
        if(first == null)
            return second;

        if(second == null)
            return first;

        Node root = null;
        Node current = root;

        int carry = 0;
        while(first != null && second != null){
            if(current != null) {
                current.next = new Node((first.data + second.data + carry) % 10);
                current = current.next;
            }
            else {
                current = new Node((first.data + second.data + carry) % 10);
            }
            carry = (first.data + second.data + carry) / 10;
            first = first.next;
            second = second.next;
        }

        Node longerList = first == null ? second : first;
        while(longerList != null){
            current.next = new Node((first.data + second.data + carry) % 10);
            current = current.next;
            longerList = longerList.next;
        }
        if(carry > 0){
            current.next = new Node(carry);
        }

        return root;
    }

    //space O(n), time O(n), recursive solution
    static Node solution5b(Node first, Node second, int carry){
        if(first == null && second == null && carry == 0)
            return null;

        Node n = new Node(carry);
        if(first != null)
            n.data += first.data;

        if(second != null)
            n.data += second.data;

        carry = n.data / 10;
        n.data = n.data % 10;

        if(first != null || second != null) {
            first = first == null ? null : first.next;
            second = second == null ? null : second.next;

            Node next = solution5b(first, second, carry);
            n.next = next;
        }

        return n;
    }

    //space O(n), time O(n)
    static boolean solution6a(Node root){
        if(root == null)
            return false;

        Node runner = root;
        Stack<Integer> dataStack = new Stack<>();

        while(runner != null){
            dataStack.push(root.data);
            root = root.next;
            runner = runner.next;
            if(runner == null) {
                dataStack.pop();
                break;
            }
            runner = runner.next;
        }

        while(root != null){
            if(root.data != dataStack.pop())
                return false;
            root = root.next;
        }

        return true;
    }

    //space O(1), time O(n)
    static Node solution7a(Node l1, Node l2){
        if(l1 == null || l2 == null)
            return null;

        Node tmpL1 = l1;
        Node tmpL2 = l2;
        int lengthL1 = 0;
        int lengthL2 = 0;

        while(tmpL1 != null){
            lengthL1++;
            tmpL1 = tmpL1.next;
        }

        while(tmpL2 != null){
            lengthL2++;
            tmpL2 = tmpL2.next;
        }

        Node longer = lengthL1 > lengthL2 ? l1 : l2;
        Node shorter = lengthL1 <= lengthL2 ? l1 : l2;

        for(int i =0; i< Math.abs(lengthL1 - lengthL2); i++){
            longer = longer.next;
        }

        while(longer != null){
            if(longer == shorter)
                return longer;

            longer = longer.next;
            shorter = shorter.next;
        }
        return null;
    }

    //space O(n), time O(n)
    static Node solution8a(Node root){
        if(root == null)
            return null;

        HashSet<Node> nodeAppeared = new HashSet<>();

        while(root != null){
            if(nodeAppeared.contains(root)) {
                return root;
            }
            else{
                nodeAppeared.add(root);
            }
            root = root.next;
        }

        return null;
    }

    //space O(1), time O(n)
    static Node solution8b(Node root){
        Node slow = root;
        Node fast = root;

        while(fast != null && fast.next != null){
            fast = fast.next.next;
            slow = slow.next;
            if(fast == slow)
                break;
        }

        if(fast == null || fast.next == null)
            return null;

        fast = root;
        while(fast != slow){
            fast = fast.next;
            slow = slow.next;
        }

        return fast;
    }
}

File: C++/CrackingTheCodingInterview/CrackingTheCodingInterview/src/LinkedLists.cpp
#include "../include/LinkedLists.h"

#include <unordered_set>
#include <memory>
#include <stack>

//space O(n), time O(n)
void solution1a(Node* head) {
	std::unordered_set<int> duplicates = std::unordered_set<int>();
	Node* slow = nullptr;
	while (head != nullptr) {
		if (duplicates.count(head->data) > 0) {
			slow->next = head->next;
		}
		else {
			duplicates.insert(head->data);
			slow = head;
		}
		head = head->next;
	}
}

//space O(1), time O(n^2)
void solution1b(Node* head) {
	if (head->next == nullptr)
		return;

	while (head != nullptr && head->next != nullptr) {
		Node* runner = head;
		while (runner->next != nullptr) {
			if (runner->next->data == head->data) {
				runner->next = runner->next->next;
			}
			else {
				runner = runner->next;
			}
		}
		head = head->next;
	}
}
//space O(1), time O(n)
Node& solution2a(Node* head, int k) {
	if (k <= 0) {
		throw std::exception("Pick K thats over 0");
	}
	Node* tmp = head;

	int counter = 0;
	while (tmp != nullptr) {
		counter++;
		tmp = tmp->next;
	}

	if (counter < k) {
		throw std::exception("Pick K thats less than length of list");
	}

	for (int i = 0; i < (counter - k); i++) {
		head = head->next;
	}

	return *head;
}
//space O(1), time O(n)
Node& solution2b(Node* head, int k) {
	if (k <= 0)
		throw std::exception("Pick K thats over 0");

	Node* runner = head;

	while (runner != nullptr && k>0) {
		k--;
		runner = runner->next;
	}

	if (k > 0) {
		throw std::exception("Pick K thats less than length of list");
	}

	while (runner != nullptr) {
		runner = runner->next;
		head = head->next;
	}

	return *head;
}

//space O(1), time O(1)
void solution3a(Node* middle) {
	if (middle->next == nullptr)
		throw std::exception("This is last node!");

	middle->data = middle->next->data;
	middle->next = middle->next->next;
}

//space O(1), time O(n)
Node& solution4a(Node* head, int k) {
	Node* smaller = nullptr;
	Node* greater = nullptr;
	Node* smallerRunner = nullptr;

	while (head != nullptr) {
		Node* next = head->next;
		if (head->data < k) {
			if (smaller == nullptr) {
				smaller = head;
				smallerRunner = smaller;
			}
			else {
				head->next = smaller;
				smaller = head;
			}
		}
		else {
			head->next = greater;
			greater = head;
		}
		head = next;
	}

	if (smallerRunner != nullptr) {
		smallerRunner->next = greater;
	}

	return smaller == nullptr ? *greater : *smaller;
}

//space O(n), time O(n)
Node& solution5a(Node* l1, Node* l2) {
	if (l1 == nullptr && l2 == nullptr) {
		throw std::exception("Both lists are empty");
	}

	if (l1 == nullptr)
		return *l2;

	if (l2 == nullptr)
		return *l1;

	Node* sum = nullptr;
	Node* runner = sum;
	int carry = 0;
	int value = 0;
	while (l1 != nullptr && l2 != nullptr) {
		value = l1->data + l2->data + carry;
		carry = value / 10;
		Node* tmp = new Node(value % 10);
		if (sum == nullptr) {
			sum = tmp;
			runner = sum;
		}
		else {
			runner->next = tmp;
			runner = runner->next;
		}
		l1 = l1->next;
		l2 = l2->next;
	}

	Node* leftOver = l1 == nullptr ? l2 : l1;
	while (leftOver != nullptr) {
		value = leftOver->data + carry;
		carry = value / 10;
		runner->next = new Node(value % 10);
		runner = runner->next;

		leftOver = leftOver->next;
	}

	if (carry > 0)
		runner->next = new Node(carry);

	return *sum;
}

//space O(1), time O(n)
bool solution6a(Node* head) {
	if (head == nullptr) {
		throw std::exception("List is empty");
	}

	std::stack<int> stackData = std::stack<int>();
	Node* fastRunner = head;

	while (fastRunner != nullptr && fastRunner->next != nullptr) {
		stackData.push(head->data);
		head = head->next;
		fastRunner = fastRunner->next->next;
	}

	if (fastRunner != nullptr)
		head = head->next;

	while (head != nullptr) {
		if (head->data != stackData.top())
			return false;
		head = head->next;
		stackData.pop();
	}

	return true;
}

//space O(1), time O(n)
Node* solution7a(Node* l1, Node* l2) {
	if (l1 == nullptr || l2 == nullptr)
		return nullptr;

	Node* tmp1 = l1;
	Node* tmp2 = l2;

	int counter1 = 0;
	int counter2 = 0;

	while (tmp1 != nullptr) {
		counter1++;
		tmp1 = tmp1->next;
	}

	while (tmp2 != nullptr) {
		counter2++;
		tmp2 = tmp2->next;
	}

	tmp1 = counter1 > counter2 ? l1 : l2;

	tmp2 = counter1 > counter2 ? l2 : l1;

	for (int i = 0; i < std::abs(counter1 - counter2); i++) {
		tmp1 = tmp1->next;
	}

	while (tmp1 != nullptr && tmp2 != nullptr) {
		if (tmp1 == tmp2)
			return tmp1;
		tmp1 = tmp1->next;
		tmp2 = tmp2->next;
	}
	
	return nullptr;
}

//space O(1), time O(n)
Node* solution8a(Node* head) {
	if (head == nullptr)
		return nullptr;

	Node* fast = head;
	Node* slow = head;

	while (fast != nullptr && fast->next != nullptr) {
		fast = fast->next->next;
		slow = slow->next;
		if (slow == fast)
			break;
	}

	if (fast == nullptr || fast->next == nullptr)
		return nullptr;

	slow = head;
	while (slow != head) {
		slow = slow->next;
		head = head->next;
	}

	return slow;
}